airbyte-integrations/bases/source-acceptance-test/unit_tests/test_backward_compatibility.py

```diff
@@ -2,976 +2,864 @@
-                            "properties": {"another_required_string": {"type": "string"}},
-                        },
-                    },
-                }
-            ),
-            does_not_raise(),
-            id="Adding an optional object with required properties should not fail.",
-        ),
-    ],
-)
-def test_new_required_field_declaration(previous_connector_spec, actual_connector_spec, expectation):
-    t = _TestSpec()
-    spec_diff = t.compute_spec_diff(actual_connector_spec, previous_connector_spec)
-    with expectation:
-        t.test_new_required_field_declaration(spec_diff)
+    def as_pytest_param(self):
+        return pytest.param(self.previous_connector_specification, self.current_connector_specification, self.should_fail, id=self.name)
 
 
-@pytest.mark.parametrize(
-    "previous_connector_spec, actual_connector_spec, expectation",
-    [
-        pytest.param(
-            ConnectorSpecification(
-                connectionSpecification={
-                    "required": ["a"],
-                }
-            ),
-            ConnectorSpecification(
-                connectionSpecification={
-                    "required": ["a", "b"],
-                }
-            ),
-            pytest.raises(AssertionError),
-            id="Top level: adding a new required property should fail.",
-        ),
-        pytest.param(
-            ConnectorSpecification(
-                connectionSpecification={
-                    "type": "object",
-                    "properties": {
-                        "my_optional_object": {
-                            "type": "object",
-                            "required": ["first_required_property"],
-                            "properties": {"first_required_property": {"type": "string"}},
-                        }
-                    },
-                }
-            ),
-            ConnectorSpecification(
-                connectionSpecification={
-                    "type": "object",
-                    "properties": {
-                        "my_optional_object": {
-                            "type": "object",
-                            "required": ["first_required_property", "second_required_property"],
-                            "properties": {
-                                "first_required_property": {"type": "string"},
-                                "second_required_property": {"type": "string"},
-                            },
-                        }
-                    },
-                }
-            ),
-            pytest.raises(AssertionError),
-            id="Nested level: adding a new required property should fail.",
-        ),
-    ],
-)
-def test_new_required_property(previous_connector_spec, actual_connector_spec, expectation):
-    t = _TestSpec()
-    spec_diff = t.compute_spec_diff(actual_connector_spec, previous_connector_spec)
-    with expectation:
-        t.test_new_required_property(spec_diff)
+FAILING_TRANSITIONS = [
```

Comment:
@pedroslopez let me know if you find this structure more readable.
I did not add a lot of new tests cases and actually removed some whose json schema was not valid (making hypothesis-jsonschema fail). (e.g  I changed `"type": []` to `"type": ["string", "string"]` etc.)

Reply:
I feel kind of neutral about it... The transitions vs specifying the params don't seem much different to me - e.g . they both define a before and after spec, a description and in some way indicate if it should fail or not, so I'm not sure what we're gaining with the Transitions.

I suppose one thing that was kind of useful in the previous implementation is that the test cases were separated by what they were testing rather than having them all together in a single list. For example, cases related to verifying if enums have been introduced were grouped together, cases related to required fields were grouped together, etc. Which may be a bit easier to think about but also makes sure we have sufficient tests for a particular problem.

Reply:
> I feel kind of neutral about it... The transitions vs specifying the params don't seem much different to me - e.g . they both define a before and after spec, a description and in some way indicate if it should fail or not, so I'm not sure what we're gaining with the Transitions.

You're right there are not much difference, but the two advantages I found are:
* You can use named arguments on data class instantiation; you can't with pytest params. E.G: We can't do pytest.param(previous_spec,currrent_spec, should_fail=True). This is minor, but I found it increases test case readability when you have to scroll through many test cases.
* If we want to serialize the spec tests case to JSON (or any other format) in the future, we could easily add a serialization function to this data class.

Let me know if you think it's overkilled. 

> I suppose one thing that was kind of useful in the previous implementation is that the test cases were separated by what they were testing rather than having them all together in a single list. For example, cases related to verifying if enums have been introduced were grouped together, cases related to required fields were grouped together, etc. Which may be a bit easier to think about but also makes sure we have sufficient tests for a particular problem.

Yes, you are right. I could split the big list into smaller scoped chunks, but on the other hand, I find it convenient for future maintenance to keep it like this. Let's say you wonder if a transition you consider failing is covered: simply add this transition to the `FAILING_TRANSITIONS` list, and you won't have to wonder to what list it should be added or create a new dedicated list.